File: apps/backend/src/services/proxy.service.ts
import { db } from '../config/database';
import { connection } from '../db/schema/connections';
import { users } from '../db/schema/users';
import { models } from '../db/schema/models';
import { eq, and, desc } from 'drizzle-orm';
import { BadRequestError } from '../utils/errors';

import { modelService } from './model.service';
import { modelPromptService } from './model-prompt.service';

export const proxyService = {

  getConnectionHeaders(conn: {
    authType: string | null;
    authValue: string | null;
    headers: string | null;
  }): Record<string, string> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };

    if (conn.authType === 'bearer' && conn.authValue) {
      headers['Authorization'] = `Bearer ${conn.authValue}`;
    } else if (conn.authType === 'api_key' && conn.authValue) {
      headers['X-API-Key'] = conn.authValue;
    }

    if (conn.headers) {
      try {
        const customHeaders = JSON.parse(conn.headers);
        Object.assign(headers, customHeaders);
      } catch { }
    }

    return headers;
  },

  async getActiveConnection(userId: string) {

    let [conn] = await db.select()
      .from(connection)
      .where(and(
        eq(connection.userId, userId),
        eq(connection.isDefault, true),
        eq(connection.isEnabled, true)
      ))
      .limit(1);

    if (!conn) {
      [conn] = await db.select()
        .from(connection)
        .where(and(
          eq(connection.userId, userId),
          eq(connection.isEnabled, true)
        ))
        .orderBy(desc(connection.priority))
        .limit(1);
    }

    if (!conn) {
      const [master] = await db.select().from(users).where(eq(users.isMaster, true)).limit(1);
      if (master) {
        [conn] = await db.select()
          .from(connection)
          .where(and(
            eq(connection.userId, master.id),
            eq(connection.isDefault, true),
            eq(connection.isEnabled, true)
          ))
          .limit(1);

        if (!conn) {
          [conn] = await db.select()
            .from(connection)
            .where(and(
              eq(connection.userId, master.id),
              eq(connection.isEnabled, true)
            ))
            .orderBy(desc(connection.priority))
            .limit(1);
        }
      }
    }

    if (!conn) {
      throw new BadRequestError('No AI provider connection configured. Please add a connection in Settings > Connections.');
    }

    return conn;
  },

  async getModels(userId: string) {

    let connections = await db.select()
      .from(connection)
      .where(and(
        eq(connection.userId, userId),
        eq(connection.isEnabled, true)
      ))
      .orderBy(desc(connection.priority));

    if (connections.length === 0) {
      const [master] = await db.select().from(users).where(eq(users.isMaster, true)).limit(1);
      if (master) {
        connections = await db.select()
          .from(connection)
          .where(and(
            eq(connection.userId, master.id),
            eq(connection.isEnabled, true)
          ))
          .orderBy(desc(connection.priority));
      }
    }

    if (connections.length === 0) {
      throw new BadRequestError('No AI provider connection configured. Please add a connection in Settings > Connections.');
    }

    const allModels: {
      id: string;
      object: string;
      created: number;
      owned_by: string;
      connection?: string;
      isEnabled?: boolean;
    }[] = [];

    for (const conn of connections) {
      try {
        const headers = this.getConnectionHeaders(conn);

        let modelsUrl = conn.url.endsWith('/')
          ? `${conn.url}v1/models`
          : `${conn.url}/v1/models`;

        let response = await fetch(modelsUrl, {
          method: 'GET',
          headers,
        });

        if (!response.ok) {
          modelsUrl = conn.url.endsWith('/')
            ? `${conn.url}models`
            : `${conn.url}/models`;

          response = await fetch(modelsUrl, {
            method: 'GET',
            headers,
          });
        }

        if (response.ok) {
          const data: any = await response.json();
          const modelsList = data.data || data.models || [];

          for (const model of modelsList) {
            const modelId = model.id || model.name;

            if (conn.modelIds) {
              const allowedIds = JSON.parse(conn.modelIds);
              if (allowedIds.length > 0 && !allowedIds.includes(modelId)) {
                continue;
              }
            }

            const finalId = conn.prefixId ? `${conn.prefixId}${modelId}` : modelId;

            allModels.push({
              id: finalId,
              object: model.object || 'model',
              created: model.created || Math.floor(Date.now() / 1000),
              owned_by: model.owned_by || conn.name,
              connection: conn.name,
            });
          }
        }
      } catch (error) {

        console.error(`Failed to fetch models from ${conn.name}:`, error);
      }
    }

    
    const userModels = await db.select()
      .from(models)
      .where(eq(models.userId, userId));

    
    const modelEnabledMap = new Map(
      userModels.map(m => [m.baseModelId || m.id, m.isEnabled])
    );

    
    const modelsWithEnabled = allModels.map(model => ({
      ...model,
      isEnabled: modelEnabledMap.has(model.id) ? modelEnabledMap.get(model.id) : true,
    }));

    return {
      object: 'list',
      data: modelsWithEnabled,
    };
  },

  async getConnectionByModel(userId: string, modelId: string) {

    let connections = await db.select()
      .from(connection)
      .where(and(
        eq(connection.userId, userId),
        eq(connection.isEnabled, true)
      ))
      .orderBy(desc(connection.priority));

    if (connections.length === 0) {
      const [master] = await db.select().from(users).where(eq(users.isMaster, true)).limit(1);
      if (master) {
        connections = await db.select()
          .from(connection)
          .where(and(
            eq(connection.userId, master.id),
            eq(connection.isEnabled, true)
          ))
          .orderBy(desc(connection.priority));
      }
    }

    if (connections.length === 0) {
      throw new BadRequestError('No AI provider connection configured.');
    }

    for (const conn of connections) {

      if (conn.prefixId && modelId.startsWith(conn.prefixId)) {
        return conn;
      }

      if (conn.modelIds) {
        try {
          const allowedIds = JSON.parse(conn.modelIds);
          if (allowedIds.includes(modelId)) {
            return conn;
          }
        } catch { }
      }
    }

    for (const conn of connections) {
      try {
        const headers = this.getConnectionHeaders(conn);
        let modelsUrl = conn.url.endsWith('/')
          ? `${conn.url}v1/models`
          : `${conn.url}/v1/models`;

        let response = await fetch(modelsUrl, { method: 'GET', headers });

        if (!response.ok) {
          modelsUrl = conn.url.endsWith('/')
            ? `${conn.url}models`
            : `${conn.url}/models`;
          response = await fetch(modelsUrl, { method: 'GET', headers });
        }

        if (response.ok) {
          const data: any = await response.json();
          const models = data.data || data.models || [];
          const modelIds = models.map((m: any) => m.id || m.name);

          const searchId = conn.prefixId ? modelId.replace(conn.prefixId, '') : modelId;
          if (modelIds.includes(modelId) || modelIds.includes(searchId)) {
            return conn;
          }
        }
      } catch { }
    }

    return this.getActiveConnection(userId);
  },

  async chatCompletions(userId: string, body: any) {

    const modelId = body.model || '';
    const promptId = body.promptId;

    if (modelId) {
      try {
        let systemPromptText = '';
        let promptSource = 'none';

        console.log('[PROXY] Starting system prompt injection for model:', modelId, 'promptId:', promptId);

        if (promptId) {
          try {
            const prompt = await modelPromptService.getModelPromptById(promptId, userId);
            if (prompt && prompt.enabled) {
              systemPromptText = prompt.prompt;
              promptSource = `custom-prompt-${prompt.name}`;
              console.log('[PROXY] ✓ Using selected custom prompt:', prompt.name, '| Length:', prompt.prompt?.length);
            } else {
              console.log('[PROXY] ✗ Selected prompt not found or disabled');
            }
          } catch (err) {
            console.error('[PROXY] ✗ Error fetching selected prompt:', err);
          }
        }

        if (!systemPromptText) {
          try {
            const modelRecord = await modelService.getOrCreateModel(modelId, userId);
            const allPrompts = await modelPromptService.getModelPrompts(modelRecord.id, userId);
            const defaultPrompt = allPrompts.find((p: any) => p.isDefault && p.enabled);
            if (defaultPrompt) {
              systemPromptText = defaultPrompt.prompt;
              promptSource = `default-prompt-${defaultPrompt.name}`;
              console.log('[PROXY] ✓ Using default model prompt:', defaultPrompt.name, '| Length:', defaultPrompt.prompt?.length);
            } else {
              console.log('[PROXY] ✗ No default prompt found');
            }
          } catch (err) {
            console.error('[PROXY] ✗ Error fetching default model prompt:', err);
          }
        }

        if (!systemPromptText) {
          const systemPromptConfig = await modelService.getSystemPrompt(modelId, userId);
          if (systemPromptConfig.enabled && systemPromptConfig.prompt) {
            systemPromptText = systemPromptConfig.prompt;
            promptSource = 'model-config';
            console.log('[PROXY] ✓ Using model config system prompt | Length:', systemPromptConfig.prompt?.length);
          } else {
            console.log('[PROXY] ✗ Model config system prompt not enabled or empty');
          }
        }

        if (systemPromptText) {
          const messages = body.messages || [];
          const hasSystemMessage = messages.length > 0 && messages[0].role === 'system';

          if (hasSystemMessage) {
            messages[0].content = systemPromptText;
            console.log('[PROXY] ✓ Replaced existing system message');
          } else {
            messages.unshift({ role: 'system', content: systemPromptText });
            console.log('[PROXY] ✓ Prepended new system message');
          }
          body.messages = messages;
          console.log('[PROXY] ✓ System prompt injected successfully | Source:', promptSource);
        } else {
          console.log('[PROXY] ✗ No system prompt to inject');
        }
      } catch (error) {

        console.error('[PROXY] Error injecting system prompt:', error);
      }
    }

    const { promptId: _, ...cleanBody } = body;

    console.log('[PROXY] Final body being sent to AI provider:', {
      model: cleanBody.model,
      messageCount: cleanBody.messages?.length,
      firstMessageRole: cleanBody.messages?.[0]?.role,
      hasSystemPrompt: cleanBody.messages?.[0]?.role === 'system',
      systemPromptPreview: cleanBody.messages?.[0]?.role === 'system'
        ? cleanBody.messages[0].content.substring(0, 100) + '...'
        : 'N/A'
    });

    const conn = modelId
      ? await this.getConnectionByModel(userId, modelId)
      : await this.getActiveConnection(userId);
    const headers = this.getConnectionHeaders(conn);

    const chatUrl = conn.url.endsWith('/')
      ? `${conn.url}v1/chat/completions`
      : `${conn.url}/v1/chat/completions`;

    const response = await fetch(chatUrl, {
      method: 'POST',
      headers,
      body: JSON.stringify(cleanBody),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new BadRequestError(`Chat completion failed: ${response.status} - ${errorText}`);
    }

    return response.json();
  },

  async chatCompletionsStream(userId: string, body: any) {

    const modelId = body.model || '';
    const promptId = body.promptId;

    if (modelId) {
      try {
        let systemPromptText = '';
        let promptSource = 'none';

        console.log('[PROXY STREAM] Starting system prompt injection for model:', modelId, 'promptId:', promptId);

        if (promptId) {
          try {
            const prompt = await modelPromptService.getModelPromptById(promptId, userId);
            if (prompt && prompt.enabled) {
              systemPromptText = prompt.prompt;
              promptSource = `custom-prompt-${prompt.name}`;
              console.log('[PROXY STREAM] ✓ Using selected custom prompt:', prompt.name, '| Length:', prompt.prompt?.length);
            } else {
              console.log('[PROXY STREAM] ✗ Selected prompt not found or disabled');
            }
          } catch (err) {
            console.error('[PROXY STREAM] ✗ Error fetching selected prompt:', err);
          }
        }

        if (!systemPromptText) {
          try {
            const modelRecord = await modelService.getOrCreateModel(modelId, userId);
            const allPrompts = await modelPromptService.getModelPrompts(modelRecord.id, userId);
            const defaultPrompt = allPrompts.find((p: any) => p.isDefault && p.enabled);
            if (defaultPrompt) {
              systemPromptText = defaultPrompt.prompt;
              promptSource = `default-prompt-${defaultPrompt.name}`;
              console.log('[PROXY STREAM] ✓ Using default model prompt:', defaultPrompt.name, '| Length:', defaultPrompt.prompt?.length);
            } else {
              console.log('[PROXY STREAM] ✗ No default prompt found');
            }
          } catch (err) {
            console.error('[PROXY STREAM] ✗ Error fetching default model prompt:', err);
          }
        }

        if (!systemPromptText) {
          const systemPromptConfig = await modelService.getSystemPrompt(modelId, userId);
          if (systemPromptConfig.enabled && systemPromptConfig.prompt) {
            systemPromptText = systemPromptConfig.prompt;
            promptSource = 'model-config';
            console.log('[PROXY STREAM] ✓ Using model config system prompt | Length:', systemPromptConfig.prompt?.length);
          } else {
            console.log('[PROXY STREAM] ✗ Model config system prompt not enabled or empty');
          }
        }

        if (systemPromptText) {
          const messages = body.messages || [];
          const hasSystemMessage = messages.length > 0 && messages[0].role === 'system';

          if (hasSystemMessage) {
            messages[0].content = systemPromptText;
            console.log('[PROXY STREAM] ✓ Replaced existing system message');
          } else {
            messages.unshift({ role: 'system', content: systemPromptText });
            console.log('[PROXY STREAM] ✓ Prepended new system message');
          }
          body.messages = messages;
          console.log('[PROXY STREAM] ✓ System prompt injected successfully | Source:', promptSource);
        } else {
          console.log('[PROXY STREAM] ✗ No system prompt to inject');
        }
      } catch (error) {

        console.error('[PROXY] Error injecting system prompt:', error);
      }
    }

    const { promptId: _, ...cleanBody } = body;

    console.log('[PROXY STREAM] Final body being sent to AI provider:', {
      model: cleanBody.model,
      messageCount: cleanBody.messages?.length,
      firstMessageRole: cleanBody.messages?.[0]?.role,
      hasSystemPrompt: cleanBody.messages?.[0]?.role === 'system',
      systemPromptPreview: cleanBody.messages?.[0]?.role === 'system'
        ? cleanBody.messages[0].content.substring(0, 100) + '...'
        : 'N/A'
    });

    const conn = modelId
      ? await this.getConnectionByModel(userId, modelId)
      : await this.getActiveConnection(userId);
    const headers = this.getConnectionHeaders(conn);

    const chatUrl = conn.url.endsWith('/')
      ? `${conn.url}v1/chat/completions`
      : `${conn.url}/v1/chat/completions`;

    const response = await fetch(chatUrl, {
      method: 'POST',
      headers,
      body: JSON.stringify({ ...cleanBody, stream: true }),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new BadRequestError(`Chat completion failed: ${response.status} - ${errorText}`);
    }

    return response;
  },

  async completions(userId: string, body: any) {
    const conn = await this.getActiveConnection(userId);
    const headers = this.getConnectionHeaders(conn);

    const completionsUrl = conn.url.endsWith('/')
      ? `${conn.url}v1/completions`
      : `${conn.url}/v1/completions`;

    const response = await fetch(completionsUrl, {
      method: 'POST',
      headers,
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new BadRequestError(`Completion failed: ${response.status} - ${errorText}`);
    }

    return response.json();
  },

  async embeddings(userId: string, body: any) {
    const conn = await this.getActiveConnection(userId);
    const headers = this.getConnectionHeaders(conn);

    const embeddingsUrl = conn.url.endsWith('/')
      ? `${conn.url}v1/embeddings`
      : `${conn.url}/v1/embeddings`;

    const response = await fetch(embeddingsUrl, {
      method: 'POST',
      headers,
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new BadRequestError(`Embeddings failed: ${response.status} - ${errorText}`);
    }

    return response.json();
  },
};

File: apps/backend-python/src/services/proxy_service.py
"""
Proxy service: forwards requests to the AI provider (OpenAI-compatible).
Matches TypeScript backend behaviour:
  - Resolves connection by model ID
  - Injects system prompt from promptId / default model prompt / model config
  - Strips unknown fields (promptId) before forwarding
"""
import httpx
from sqlalchemy.orm import Session
from src.db.models import Connection, Model, ModelPrompt, User
from src.utils.errors import BadRequestError

_DEFAULT_TIMEOUT = 120.0


# ---------------------------------------------------------------------------
# Connection helpers
# ---------------------------------------------------------------------------

def _get_default_connection(db: Session, user_id: str) -> Connection:
    """Return the default/highest-priority enabled connection for user."""
    conn = (
        db.query(Connection)
        .filter(Connection.user_id == user_id, Connection.is_enabled == True)
        .order_by(Connection.is_default.desc(), Connection.priority.desc())
        .first()
    )
    if conn:
        return conn

    # Fallback: master user's connection
    master = db.query(User).filter(User.is_master == True).first()
    if master:
        conn = (
            db.query(Connection)
            .filter(Connection.user_id == master.id, Connection.is_enabled == True)
            .order_by(Connection.is_default.desc(), Connection.priority.desc())
            .first()
        )
        if conn:
            return conn

    raise BadRequestError("No AI provider connection configured")


def _get_connection_by_model(db: Session, user_id: str, model_id: str) -> Connection:
    """Return the most appropriate connection for a given model ID."""
    connections = (
        db.query(Connection)
        .filter(Connection.user_id == user_id, Connection.is_enabled == True)
        .order_by(Connection.priority.desc())
        .all()
    )

    if not connections:
        master = db.query(User).filter(User.is_master == True).first()
        if master:
            connections = (
                db.query(Connection)
                .filter(Connection.user_id == master.id, Connection.is_enabled == True)
                .order_by(Connection.priority.desc())
                .all()
            )

    for conn in connections:
        if conn.prefix_id and model_id.startswith(conn.prefix_id):
            return conn

    return _get_default_connection(db, user_id)


def _build_headers(conn: Connection) -> dict:
    headers = {"Content-Type": "application/json"}
    if conn.auth_type in ("bearer", "api_key") and conn.auth_value:
        headers["Authorization"] = f"Bearer {conn.auth_value}"
    return headers


def _chat_url(conn: Connection) -> str:
    return f"{conn.url.rstrip('/')}/v1/chat/completions"


# ---------------------------------------------------------------------------
# System prompt injection (mirrors TypeScript proxyService)
# ---------------------------------------------------------------------------

def _get_or_create_model(db: Session, model_id: str, user_id: str) -> Model:
    """Find existing model by base_model_id or create a lightweight record."""
    from datetime import datetime
    from nanoid import generate

    m = db.query(Model).filter(
        Model.base_model_id == model_id, Model.user_id == user_id
    ).first()
    if m:
        return m

    # Create placeholder
    m = Model(
        id=generate(),
        user_id=user_id,
        base_model_id=model_id,
        name=model_id,
        is_enabled=True,
        created_at=datetime.utcnow(),
        updated_at=datetime.utcnow(),
    )
    db.add(m)
    db.commit()
    db.refresh(m)
    return m


def _inject_system_prompt(db: Session, user_id: str, body: dict) -> dict:
    """
    1. Strip `promptId` from body.
    2. Resolve system prompt text (promptId → default model prompt → model config).
    3. Inject as first system message.
    Returns a clean copy of body ready to forward.
    """
    body = dict(body)  # shallow copy
    model_id = body.get("model", "")
    prompt_id = body.pop("promptId", None)  # always strip

    system_prompt_text = ""

    # 1. Explicit promptId
    if prompt_id:
        try:
            mp = db.query(ModelPrompt).filter(
                ModelPrompt.id == prompt_id, ModelPrompt.user_id == user_id
            ).first()
            if mp and mp.enabled:
                system_prompt_text = mp.prompt or mp.content or ""
                print(f"[PROXY] ✓ Using selected prompt: {mp.name}")
            else:
                print("[PROXY] ✗ promptId not found or disabled")
        except Exception as e:
            print(f"[PROXY] ✗ Error fetching promptId: {e}")

    # 2. Default model prompt
    if not system_prompt_text and model_id:
        try:
            model_record = _get_or_create_model(db, model_id, user_id)
            default_mp = db.query(ModelPrompt).filter(
                ModelPrompt.model_id == model_record.id,
                ModelPrompt.user_id == user_id,
                ModelPrompt.is_default == True,
                ModelPrompt.enabled == True,
            ).first()
            if default_mp:
                system_prompt_text = default_mp.prompt or default_mp.content or ""
                print(f"[PROXY] ✓ Using default model prompt: {default_mp.name}")
            else:
                print("[PROXY] ✗ No default model prompt")
        except Exception as e:
            print(f"[PROXY] ✗ Error fetching default prompt: {e}")

    # 3. Model config system prompt (legacy)
    if not system_prompt_text and model_id:
        try:
            model_record = _get_or_create_model(db, model_id, user_id)
            legacy_mp = db.query(ModelPrompt).filter(
                ModelPrompt.model_id == model_record.id,
                ModelPrompt.user_id == user_id,
            ).first()
            if legacy_mp:
                text = legacy_mp.prompt or legacy_mp.content or ""
                if text:
                    system_prompt_text = text
                    print("[PROXY] ✓ Using legacy model config prompt")
        except Exception as e:
            print(f"[PROXY] ✗ Error fetching legacy prompt: {e}")

    # Inject
    if system_prompt_text:
        messages = list(body.get("messages", []))
        if messages and messages[0].get("role") == "system":
            messages[0] = {"role": "system", "content": system_prompt_text}
            print("[PROXY] ✓ Replaced existing system message")
        else:
            messages.insert(0, {"role": "system", "content": system_prompt_text})
            print("[PROXY] ✓ Prepended new system message")
        body["messages"] = messages
    else:
        print("[PROXY] ✗ No system prompt to inject")

    return body


# ---------------------------------------------------------------------------
# Service class
# ---------------------------------------------------------------------------

class ProxyService:
    def get_models(self, db: Session, user_id: str) -> dict:
        conn = _get_default_connection(db, user_id)
        resp = httpx.get(
            f"{conn.url.rstrip('/')}/v1/models",
            headers=_build_headers(conn),
            timeout=_DEFAULT_TIMEOUT,
        )
        resp.raise_for_status()
        return resp.json()

    def chat_completions(self, db: Session, user_id: str, body: dict) -> dict:
        clean_body = _inject_system_prompt(db, user_id, body)
        model_id = clean_body.get("model", "")
        conn = _get_connection_by_model(db, user_id, model_id) if model_id else _get_default_connection(db, user_id)
        resp = httpx.post(
            _chat_url(conn),
            json=clean_body,
            headers=_build_headers(conn),
            timeout=_DEFAULT_TIMEOUT,
        )
        resp.raise_for_status()
        return resp.json()

    def completions(self, db: Session, user_id: str, body: dict) -> dict:
        conn = _get_default_connection(db, user_id)
        resp = httpx.post(
            f"{conn.url.rstrip('/')}/v1/completions",
            json=body,
            headers=_build_headers(conn),
            timeout=_DEFAULT_TIMEOUT,
        )
        resp.raise_for_status()
        return resp.json()

    def embeddings(self, db: Session, user_id: str, body: dict) -> dict:
        conn = _get_default_connection(db, user_id)
        resp = httpx.post(
            f"{conn.url.rstrip('/')}/v1/embeddings",
            json=body,
            headers=_build_headers(conn),
            timeout=_DEFAULT_TIMEOUT,
        )
        resp.raise_for_status()
        return resp.json()


proxy_service = ProxyService()
